src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12912,12 +12912,35 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
         private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToken, SeparatedSyntaxListBuilder<T> list, SyntaxKind expected)
             where T : CSharpSyntaxNode
         {
+            bool openBraceIsMissing = startToken.IsMissing;
             return this.SkipBadSeparatedListTokensWithExpectedKind(ref startToken, list,
                 p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleExpression(),
-                p => p.CurrentToken.Kind == SyntaxKind.CloseBraceToken || p.IsTerminator(),
+                p => p.IsEndOfInitializer(openBraceIsMissing),
                 expected);
         }
 
+        private bool IsEndOfInitializer(bool openBraceIsMissing)
+        {
+            if (!openBraceIsMissing && CurrentToken.Kind == SyntaxKind.SemicolonToken)
+            {
+                // When encountering a semicolon in an itializer body, the user may have mistakenly intended it as a separator.
+
+                // Seek ahead and check if the semicolon is being followed by:
+                // 1. a closing brace - the semicolon is not the end, the closing brace is the end
+                // 2. another variable intializer - not the end yet
+                var resetPoint = GetResetPoint();
+                EatToken();
+                bool initializerContinues = CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsPossibleVariableInitializer();
+                Reset(ref resetPoint);
+                Release(ref resetPoint);
+                return !initializerContinues;
+            }
+            // the current token ends the initializer when it is:
+            // 1. a close brace
+            // 2. another type of terminator
+            return CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsTerminator();
```

Comment:
nit: this is the trivial logic.  my preference with code would be to invert htings so we have:

```c#
if (openBraceIsMissing || CurrentToken.Kind != SyntaxKind.SemicolonToken)
    return trivialLogic;
    
doComplexLogic();
```

it just reads more easily.  thanks!

Reply:
✔️